### PR TITLE
[DT-118][risk-no] change generated BQ SQL for any concept set that contains questions from PFHH survey

### PR DESCRIPTION
**Pull request by freemabd**
change generated BQ SQL for any concept set that contains questions from PFHH survey

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java [hunk 1]
```diff
@@ -440,6 +440,40 @@ List<DbSurveyVersion> findSurveyVersionByQuestionConceptIdAndAnswerConceptId(
...
+              + "      from cb_criteria "
+              + "      where concept_id in (1740639) "
+              + "      and domain_id = 'SURVEY' "
+              + "      ) a on (c.path like CONCAT('%', a.id, '.%')) "
+              + "where domain_id = 'SURVEY' "
+              + "and type = 'PPI' "
+              + "and subtype = 'QUESTION' "
+              + "and concept_id in (:conceptIds)",
+      nativeQuery = true)
+  List<Long> findPFHHSurveyQuestionIds(@Param("conceptIds") List<Long> conceptIds);
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java:458]: This is the lookup for find any questions that exist in the PFHH survey(1740639)

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java [hunk 2]
```diff
@@ -440,6 +440,40 @@ List<DbSurveyVersion> findSurveyVersionByQuestionConceptIdAndAnswerConceptId(
...
+              + "      from cb_criteria "
+              + "      where concept_id in (1740639) "
+              + "      and domain_id = 'SURVEY' "
+              + "      ) a on (c.path like CONCAT('%', a.id, '.%')) "
+              + "where domain_id = 'SURVEY' "
+              + "and type = 'PPI' "
+              + "and subtype = 'ANSWER' "
+              + "and concept_id in (:conceptIds)",
+      nativeQuery = true)
+  List<Long> findPFHHSurveyAnswerIds(@Param("conceptIds") List<Long> conceptIds);
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java:475]: This is the lookup for answers in the PFHH survey(1740639)

File: api/src/bigquerytest/resources/bigquery/cbdata/ds_survey_data.json [hunk 3]
```diff
@@ -7,5 +7,14 @@
     "QUESTION": "Question",
     "ANSWER_CONCEPT_ID": 1,
     "ANSWER": "NO"
+  },
+  {
+    "PERSON_ID": "1",
+    "SURVEY_DATETIME": "2020-01-31 00:00:00 UTC",
+    "SURVEY": "The Basics",
```
Comment (chenchals) [inline api/src/bigquerytest/resources/bigquery/cbdata/ds_survey_data.json:14]: Should this survey name be: "Personal and Family Health History" ?

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java (hunk 1 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java:458]: so the conceptIds are always question_concept_ids right? and this query is against MySQL and not BQ

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java (hunk 2 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java:475]: so the conceptIds are always question_concept_ids right? and this query is against MySQL and not BQ

File: api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java [hunk 4]
```diff
@@ -107,12 +112,15 @@ public class DataSetControllerBQTest extends BigQueryBaseTest {
 
   @Autowired private AccessTierDao accessTierDao;
   @Autowired private BigQueryService bigQueryService;
+
```
Comment (chenchals) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java:115]: delete empty line

File: api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java [hunk 5]
```diff
@@ -107,12 +112,15 @@ public class DataSetControllerBQTest extends BigQueryBaseTest {
...
+  @Autowired private CBCriteriaDao cbCriteriaDao;
   @Autowired private CdrVersionDao cdrVersionDao;
   @Autowired private CdrVersionService cdrVersionService;
+
```
Comment (chenchals) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java:119]: delete empty line

File: api/src/bigquerytest/resources/bigquery/cbdata/ds_survey_data.json (hunk 3 above)
Comment (freemabd) [reply api/src/bigquerytest/resources/bigquery/cbdata/ds_survey_data.json]: Done

File: api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java [hunk 6]
```diff
@@ -146,6 +154,8 @@ public class DataSetControllerBQTest extends BigQueryBaseTest {
   private DbConceptSet dbConditionConceptSetForValues2;
   private DbConceptSet dbProcedureConceptSet;
   private DbConceptSet dbMeasurementConceptSet;
+
```
Comment (chenchals) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java:156]: delete empty line

File: api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java (hunk 4 above)
Comment (freemabd) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java]: Done

File: api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java (hunk 5 above)
Comment (freemabd) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java]: Done.

File: api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java (hunk 6 above)
Comment (freemabd) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/DataSetControllerBQTest.java]: Done

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java [hunk 7]
```diff
@@ -230,6 +231,8 @@ public Map<String, QueryParameterValue> getNamedParameterValues() {
   }
 
   private final BigQueryService bigQueryService;
+
```
Comment (chenchals) [inline api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:234]: delete empty line

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java [hunk 8]
```diff
@@ -366,13 +371,30 @@ public TableResult previewBigQueryJobConfig(DataSetPreviewRequest request) {
...
+          List<Long> pfhhSurveyQuestionIds = findPFHHSurveyQuestionIds(questionConceptIds);
+          if (!pfhhSurveyQuestionIds.isEmpty()) {
+            // need to filter out PFHH survey questions for other survey questions
+            dbConceptSetConceptIds =
+                dbConceptSetConceptIds.stream()
+                    .filter(cid -> !pfhhSurveyQuestionIds.contains(cid.getConceptId()))
+                    .collect(Collectors.toList());
+            // find all answers for the questions
+            dbCriteriaAnswerIds = findPFHHSurveyAnswerIds(pfhhSurveyQuestionIds);
+          }
```
Comment (chenchals) [inline api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:396]: Why is this block under `case PHYSICAL_MEASUREMENT_CSS:` and not under `case SURVEY:` or under default?
I do not understand

Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:396]: Notice the code, both PHYSICAL_MEASUREMENT_CSS and SURVEY fall under this section:
Screen Shot 2022-11-30 at 9 39 43 AM

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java [hunk 9]
```diff
@@ -717,56 +748,80 @@ private Optional<String> buildConceptIdSqlInClause(
...
+          queryBuilder.append(" OR ");
         }
-        return Optional.of("(" + queryBuilder + ")");
+        queryBuilder.append(
+            "answer_concept_id IN (@answerConceptIds)"
+                .replaceAll("@answerConceptIds", answerConceptIds));
       }
+      return Optional.of("(" + queryBuilder + ")");
     }
   }
```
Comment (chenchals) [inline api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:821]: L#745-820 private Optional<String> buildConceptIdSqlInClause needs some refactor very hard to follow... just a suggestion- 75 lines of different if-else blocks...

**Review by chenchals — CHANGES_REQUESTED** [review]
Wow thats a lot of code change. Thank you.

File: api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java (hunk 9 above)
Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java:821]: Refactor is not a priority right now. Please create a ticket for a refactor if you think you can make this better.

**Review by chenchals — APPROVED** [review]
Thank you!
